omegaconf/omegaconf.py

```diff
@@ -830,10 +848,11 @@ def _maybe_wrap(
     value: Any,
     is_optional: bool,
     parent: Optional[BaseContainer],
+    use_type: bool = False,
 ) -> Node:
     # if already a node, update key and parent and return as is.
     # NOTE: that this mutate the input node!
-    if isinstance(value, Node):
+    if isinstance(value, Node) and ref_type is Any:
```

Comment:
So, this is the change I'm most concerned about. Until now `_node_wrap()` was never called (I think) with `value` already being a Node, and I'm a bit worried of what might happen if we allow this (especially if the code doesn't explicitly restricts this to unions).

I don't have a good solution yet, I'm still trying to figure out exactly how and when `_maybe_wrap()` is being used thoughout the codebase. But if there was a way to make sure that the changes required to support Union won't have unintended side effects elsewhere, I would be more confident in letting them in.

I may not have time to look more closely at it in the next few days, sorry. @omry feel free to chime in :)

Reply:
Wrapping of Nodes is causing me trouble.

Reply:
> Wrapping of Nodes is causing me trouble.

Yeah, honestly, it seems brittle to me, because the rest of the code wasn't written with that in mind. There are a few weird things going on that I can see, and probably more that I can't.

I submitted a PR to your branch in https://github.com/pereman2/omegaconf/pull/1 with a few things I believe make it at least a bit better. For some of my changes I struggled to write a meaningful test as I got deep into some rabbit holes, so eventually I gave up and decided that just "looking better" would be good enough ;)
(in particular, now `_node_wrap()` will return a `UnionNode` as soon as `type_` is a union type, even if called from outside of `UnionNode`)

I doubt we'll manage to get to something 100% satisfying without a deep re-thinking of some OmegaConf internals (or completely changing the approach taken here), so as far as I'm concerned I'm ok with accepting this PR (with my suggested changes), unless @omry thinks otherwise.

Reply:
IIRC, The reason maybe_wrap exists is exactly to avoid that.
wrap this object if it's not already a node. so generally I think _node_wrap should not be used on nodes (and should probably assert that).

I am going to look at the code after going through the comment and form a more informed opinion.

Reply:
> I think _node_wrap should not be used on nodes (and should probably assert that).

Just to give some context, currently we need it because otherwise with something like:
```python
 x: Union[int, DictConfig] = field(default_factory=lambda: DictConfig({}))
```
the type of node `x` is `DictConfig` instead of `UnionNode`, and this may break things later when we try to assign other values to `x`.

`UnionNode` is an odd beast because it's a `ValueNode` (not a `Container`) but it wraps another node -- something that didn't exist before.